File: src/ios/YozioPlugin.m
#import "Yozio.h"
#import "YozioPlugin.h"
#import <Cordova/CDVPlugin.h>
#import <Cordova/CDVCommandDelegateImpl.h>

@implementation YozioPlugin

CDVCommandDelegateImpl *newInstallCommandDelegateHandler = nil;
CDVInvokedUrlCommand *newInstallCommandHandler = nil;
CDVCommandDelegateImpl *deeplinkCommandDelegateHandler = nil;
CDVInvokedUrlCommand *deeplinkCommandHandler = nil;

- (void)onNewInstallMetadataCallback:(CDVInvokedUrlCommand *)command
{
    newInstallCommandDelegateHandler = self.commandDelegate;
    newInstallCommandHandler = command;
}

// called when Yozio SDK initialize, pass metadata back to javascript interface
+ (void)onYozioSDKNewInstallMetadataCallback:(NSDictionary *)metadata
{
    if (newInstallCommandHandler != nil && newInstallCommandHandler != nil) {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:metadata];
        [newInstallCommandDelegateHandler sendPluginResult:pluginResult callbackId:newInstallCommandHandler.callbackId];
    }
}

- (void)onDeeplinkMetadataCallback:(CDVInvokedUrlCommand *)command
{
    deeplinkCommandDelegateHandler = self.commandDelegate;
    deeplinkCommandHandler = command;
}

// called when app opened through deeplink, pass metadata back to javascript interface
+ (void)onYozioSDKDeeplinkMetadataCallback:(NSDictionary *)metadata
{
    if (deeplinkCommandHandler != nil && deeplinkCommandHandler != nil) {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:metadata];
        [deeplinkCommandDelegateHandler sendPluginResult:pluginResult callbackId:deeplinkCommandHandler.callbackId];
    }
}

- (void)getInstallMetadata:(CDVInvokedUrlCommand *)command
{
    NSDictionary *metadata = [Yozio getNewInstallMetaDataAsHash];
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:metadata];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getLastDeeplinkMetadata:(CDVInvokedUrlCommand *)command
{
    NSDictionary *metadata = [Yozio getLastDeeplinkMetaDataAsHash];
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:metadata];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)trackSignup:(CDVInvokedUrlCommand *)command
{
	// Delegate to the Yozio SDK.
	[self.commandDelegate runInBackground:^{
		[Yozio trackSignup];
	}];

	CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)trackPayment:(CDVInvokedUrlCommand *)command {

	// Ensure we have the correct number of arguments.
	if ([command.arguments count] != 1) {
		CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"An amount is required."];
		[self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
		return;
	}

	// Obtain the arguments.
	NSNumber* amount = [command.arguments objectAtIndex:0];

	// Validate the arguments.
	if (!amount) {
		CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"An amount is required."];
		[self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
		return;
	}

	// Delegate to the Yozio SDK.
	[self.commandDelegate runInBackground:^{
		[Yozio trackPayment:[amount doubleValue]];
	}];

	CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
	[self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];

}

- (void)trackCustomEvent:(CDVInvokedUrlCommand *)command
{
    if ([command.arguments count] != 2) {
        CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"An event name and amount are required."];
        [self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        return;
    }
    
    // Obtain the arguments.
    NSString* eventName = [command.arguments objectAtIndex:0];
    NSNumber* value = [command.arguments objectAtIndex:1];
    
    // Validate the arguments.
    if (!eventName) {
        CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"An event name is required."];
        [self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        return;
    }
    
    if (!value) {
        CDVPluginResult *res = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"A value is required."];
        [self.commandDelegate sendPluginResult:res callbackId:command.callbackId];
        return;
    }
    
    // Delegate to the Yozio SDK.
    [self.commandDelegate runInBackground:^{
        [Yozio trackCustomEventWithName:eventName andValue:[value doubleValue]];
    }];
    
    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    
}

@end

File: src/android/com/yozio/YozioPlugin.java
package com.yozio;

import android.content.Intent;

import com.yozio.android.Yozio;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

public class YozioPlugin extends CordovaPlugin {

	private static CallbackContext newInstallCallbackContext = null;
	private static CallbackContext deeplinkCallbackContext = null;

    private static boolean isOpenedByNewInstallCallback = false;

    public static void setIsOpenedByNewInstallCallback(boolean value) {
        isOpenedByNewInstallCallback = value;
    }

	@Override
	protected void pluginInitialize() {

		Yozio.YOZIO_ENABLE_LOGGING = true;
		Yozio.YOZIO_READ_TIMEOUT = 7000;

		Yozio.initialize(cordova.getActivity());
	}

	@Override
	public boolean execute(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
		if ("onNewInstallMetadataCallback".equals(action)) {

			YozioPlugin.newInstallCallbackContext = callbackContext;
			return true;

		} else if ("onDeeplinkMetadataCallback".equals(action)) {

			YozioPlugin.deeplinkCallbackContext = callbackContext;
			return true;

		} else if ("getInstallMetadata".equals(action)) {

			this.getInstallMetadata(callbackContext);
			return true;

		} else if ("getLastDeeplinkMetadata".equals(action)) {

			this.getLastDeeplinkMetadata(callbackContext);
			return true;

		} else if ("trackSignup".equals(action)) {

			cordova.getThreadPool().execute(new Runnable() {
				public void run() {
					YozioPlugin.this.trackSignup(callbackContext);
				}
			});
			return true;

		} else if ("trackPayment".equals(action)) {

			cordova.getThreadPool().execute(new Runnable() {
				public void run() {
					YozioPlugin.this.trackPayment(args, callbackContext);
				}
			});
			return true;

		} else if ("trackCustomEvent".equals(action)) {

			cordova.getThreadPool().execute(new Runnable() {
				public void run() {
					YozioPlugin.this.trackCustomEvent(args, callbackContext);
				}
			});

			return true;
		}
		return false;  // Returning false results in a "MethodNotFound" error.
	}

	public static void onNewInstallMetadataCallback(HashMap<String, Object> metadata) {
		if (newInstallCallbackContext != null) {
			JSONObject metadataObj = new JSONObject(metadata);
			PluginResult result = new PluginResult(PluginResult.Status.OK, metadataObj);
			result.setKeepCallback(true);
			newInstallCallbackContext.sendPluginResult(result);
		}
	}

	@Override
	public void onNewIntent(Intent intent) {
		super.onNewIntent(intent);

		HashMap<String, Object> metadata = Yozio.getMetaData(intent);
		if (!isOpenedByNewInstallCallback) {
			isOpenedByNewInstallCallback = false;
			this.onDeeplinkMetadataCallback(metadata);
		}
	}

	private void onDeeplinkMetadataCallback(HashMap<String, Object> metadata) {

		if (deeplinkCallbackContext != null) {

			JSONObject metadataObj = new JSONObject(metadata);
			PluginResult result = new PluginResult(PluginResult.Status.OK, metadataObj);
			result.setKeepCallback(true);
			deeplinkCallbackContext.sendPluginResult(result);
		}
	}

	private void getInstallMetadata(final CallbackContext callbackContext) {

		HashMap<String, Object> metadata = Yozio.getInstallMetaDataAsHash(cordova.getActivity().getApplicationContext());
		callbackContext.success(new JSONObject(metadata));
	}

	private void getLastDeeplinkMetadata(final CallbackContext callbackContext) {

		HashMap<String, Object> metadata = Yozio.getLastDeeplinkMetaDataAsHash(cordova.getActivity().getApplicationContext());
		callbackContext.success(new JSONObject(metadata));
	}

	private void trackSignup(final CallbackContext callbackContext) {

		Yozio.trackSignUp(cordova.getActivity().getApplicationContext());
		callbackContext.success();
	}

	private void trackPayment(JSONArray args, final CallbackContext callbackContext) {

		try {
			// Ensure we have the correct number of arguments.
			if (args.length() != 1) {
				callbackContext.error("An amount is required.");
				return;
			}

			// Obtain the arguments.
			double amount = args.getDouble(0);

			Yozio.trackPayment(cordova.getActivity().getApplicationContext(), amount);

			callbackContext.success();

		} catch (JSONException exception) {
			callbackContext.error("YozioPlugin uncaught exception: " + exception.getMessage());
		}
	}

	private void trackCustomEvent(JSONArray args, final CallbackContext callbackContext) {
		try {
			// Ensure we have the correct number of arguments.
			if (args.length() != 2) {
				callbackContext.error("An event name and amount are required.");
				return;
			}

			// Obtain the arguments.
			String eventName = args.getString(0);
			double value = args.getDouble(1);

			// Validate the arguments.

			if (eventName == null || eventName.equals("")) {
				callbackContext.error("An event name is required.");
			}

			Yozio.trackCustomEvent(cordova.getActivity().getApplicationContext(), eventName, value);

			callbackContext.success();

		} catch (JSONException exception) {
				callbackContext.error("YozioPlugin uncaught exception: " + exception.getMessage());
		}
	}

}
